src/main/java/com/google/sps/servlets/MapMarker.java

```diff
@@ -85,29 +88,52 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     Gson gson = new Gson();
     Marker newMarker = gson.fromJson(requestBody, Marker.class);
 
-    // Create new entity.
-    Entity markerEntity = new Entity("Marker");
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
-    markerEntity.setProperty("lat", newMarker.getLat());
-    markerEntity.setProperty("lng", newMarker.getLng());
+    // To store new marker.
+    if (newMarker.getId() == -1) {
+      // Create new entity.
+      Entity markerEntity = new Entity("Marker");
 
-    markerEntity.setProperty("visitHour", newMarker.getVisitHour());
-    markerEntity.setProperty("visitMinute", newMarker.getVisitMinute());
+      markerEntity.setProperty("lat", newMarker.getLat());
+      markerEntity.setProperty("lng", newMarker.getLng());
 
-    markerEntity.setProperty("leaveHour", newMarker.getLeaveHour());
-    markerEntity.setProperty("leaveMinute", newMarker.getLeaveMinute());
+      markerEntity.setProperty("visitHour", newMarker.getVisitHour());
+      markerEntity.setProperty("visitMinute", newMarker.getVisitMinute());
 
-    markerEntity.setProperty("name", newMarker.getMarkerName());
+      markerEntity.setProperty("leaveHour", newMarker.getLeaveHour());
+      markerEntity.setProperty("leaveMinute", newMarker.getLeaveMinute());
 
-    // Store new marker.
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    datastore.put(markerEntity);
+      markerEntity.setProperty("name", newMarker.getMarkerName());
 
-    Long id = markerEntity.getKey().getId();
-    gson = new Gson();
-    String json = gson.toJson(id);
+      // Store new marker.
+      datastore.put(markerEntity);
 
-    response.setContentType("application/json;");
-    response.getWriter().println(json);
+      Long id = markerEntity.getKey().getId();
+      gson = new Gson();
+      String json = gson.toJson(id);
+
+      response.setContentType("application/json;");
+      response.getWriter().println(json);
+      return;
+    }
+
+    // To update settings on an existing  markers.
+    Long id = newMarker.getId();
+
+    try {
+      Entity markerEntity = datastore.get(KeyFactory.createKey("Marker", id));
+
+      markerEntity.setProperty("visitHour", newMarker.getVisitHour());
+      markerEntity.setProperty("visitMinute", newMarker.getVisitMinute());
+
+      markerEntity.setProperty("leaveHour", newMarker.getLeaveHour());
+      markerEntity.setProperty("leaveMinute", newMarker.getLeaveMinute());
+
+      markerEntity.setProperty("name", newMarker.getMarkerName());
+      datastore.put(markerEntity);
+    } catch(Exception e) {
+
```

Comment:
I think this "two cases" division is a little strange, but this will probably be solved in a future PR when all the markers data will be actually stored just if the entire route is submitted

Reply:
I could create new servlet to update markers, but I thought that it could look a bit messy. Also I thought that to store and to update markers you should use same functions (setproperty and datastore.put) so that the only difference among them is that one use existing ID to find marker and the other one create new one.

Reply:
This can be fixed as part of a follow-up PR. There is some common code between the two. A good option is usually to remove duplication. If you have duplicate code you have to remember to make the same changes in multiple places.

The way I would approach this is:

- if the id is -1
  - create a new entity and set the long and lat
- else
  - load the entity from the data store
- set the other properties
- store the entity
- get the id from the entity
- return the id as JSON to the client.

In this way the code only divergese where needed, but there is no duplication in the rest of the code. This is a general programming principle called DRY (Don't Repeat Yourself).